R-package/R/xgb.DMatrix.R

```diff
@@ -195,10 +300,491 @@ xgb.DMatrix <- function(
+#' to the custom user-supplied callback function that forms part of the data iterator.
+#'
+#' For more information and for example usage, see the documentation for \link{xgb.ExternalDMatrix}.
+#' @return Always returns value 'TRUE'.
+#' @seealso \link{xgb.DataIter}, \link{xgb.ExternalDMatrix}.
+#' @export
+xgb.ProxyDMatrix <- function(
+  proxy_handle,
+  data,
+  label = NULL,
+  weight = NULL,
+  base_margin = NULL,
+  feature_names = colnames(data),
+  feature_types = NULL,
+  group = NULL,
+  qid = NULL,
+  label_lower_bound = NULL,
+  label_upper_bound = NULL,
+  feature_weights = NULL,
+  enable_categorical = FALSE
+) {
+  if (!inherits(proxy_handle, "xgb.ProxyDMatrix")) {
+    stop("'proxy_handle' is invalid. Can only take proxy handles supplied by 'xgb.DataIter'.")
+  }
+  if (.Call(XGCheckNullPtr_R, proxy_handle)) {
+    stop("'proxy_handle' is invalid. Was the handle serialized?")
+  }
+  if (!is.null(group) && !is.null(qid)) {
+    stop("Either one of 'group' or 'qid' should be NULL")
+  }
+
+  if (is.data.frame(data)) {
+    tmp <- .process.df.for.dmatrix(data, enable_categorical, feature_types)
+    feature_types <- tmp$feature_types
+    .Call(XGProxyDMatrixSetDataColumnar_R, proxy_handle, tmp$lst)
+    rm(tmp)
+  } else if (is.matrix(data)) {
+    .Call(XGProxyDMatrixSetDataDense_R, proxy_handle, data)
+  } else if (inherits(data, "dgRMatrix")) {
+    lst <- list(p = data@p, j = data@j, x = data@x, ncol = ncol(data))
+    .Call(XGProxyDMatrixSetDataCSR_R, proxy_handle, lst)
+  } else {
+    stop("'data' has unsupported type.")
+  }
+
+  .set.dmatrix.fields(
+    dmat = proxy_handle,
+    label = label,
+    weight = weight,
+    base_margin = base_margin,
+    feature_names = feature_names,
+    feature_types = feature_types,
+    group = group,
+    qid = qid,
+    label_lower_bound = label_lower_bound,
+    label_upper_bound = label_upper_bound,
+    feature_weights = feature_weights
+  )
+
+  return(TRUE)
+}
+
+#' @title DMatrix from External Data
+#' @description Create an xgboost 'DMatrix' or 'QuantileDMatrix' object from external data
+#' supplied by an \link{xgb.DataIter} object, potentially passed in batches from a
+#' bigger set that might not fit entirely in memory.
+#'
+#' Might be created as either a regular 'DMatrix', or as a 'QuantileDMatrix'.\itemize{
+#' \item If created as a regular 'DMatrix', the data is accessed on-demand as needed, multiple
+#' times, without being concatenated (but note that fields like 'label' \bold{will} be concatenated
+#' from multiple calls to the data iterator).
+#' \item If created as a 'QuantileDMatrix', the quantized representation of the full data will
+#' be created in memory, concatenated from multiple calls to the data iterator. The quantized
+#' version is typically lighter than the original data, so there might be cases in which this
+#' representation could potentially fit in memory even if the full data doesn't.
+#' }
```

Comment:
Yes, it does sound quite confusing, but I don't think the python interface is the best reference here.

If I understand it correctly, the issue here is that there's two routes in this PR which give the same object type:
* `xgb.QuantileDMatrix` + in-memory data.
* `xgb.ExternalDMatrix` + `xgb.DataIter` + `as_quantile_dmatrix=TRUE`.

But if this were to be implemented the same way as in python (accepting `xgb.DataIter` as `data` instead of having a dedicated class for it), then there would be:
* A function `xgb.DMatrix` which could still produce different output types depending on its inputs (regular DMatrix vs. the 'ExternalDMatrix' introduced here).
* A parameter `missing` which is auto-modified for integer inputs in some cases but not others.
* Multiple parameters that can contradict each other (e.g. feature names).

If the idea is to avoid inconsistencies, then I think the best option would be to have only two functions:
* `xgb.DMatrix`, taking only in-memory data, and accepting a parameter `as_quantile_dmatrix` in order to produce class `xgb.QuantileDMatrix` as output.
* `xgb.ExternalDMatrix`, taking only `xgb.DataIter` as input, and also accepting a parameter `as_quantile_dmatrix`.

Either that, or 4 functions after an 'ExternalQuantileDMatrix' gets implemented.

But I can also guess that:
* There will be much fewer people using `xgb.ExternalDMatrix` than the others.
* People using `xgb.ExternalDMatrix` will likely be the more advanced users.
* People using `xgb.ExternalDMatrix` will probably be more worried about getting the data iterators right and wondering about their usage than about other DMatrix properties.

Which is why I chose this kind of separation.

If there was an ExternalQuantileDMatrix type implemented, perhaps it'd make more sense to have 4 possible constructors, but I'm not sure about what's the best way to define these constructors and classes with the current behavior.

Reply:
> but I don't think the python interface is the best reference here

I agree. For background, the interface of Python is mostly caused by the use of text file input like `data="path/to/csv"`, which has the same problem with the iterator with confusing parameters.

> A parameter missing which is auto-modified for integer inputs in some cases but not others.

Not entirely sure what this means. I think as long as a missing value is accepted, integer will be considered as float.

Thank you for listing all the possible options, and apologies for the slow replies lately, will be more active starting tomorrow.

In general, I would still like to differentiate `DMatrix` and `QuantileDMatrix` since they have different functionalities and accept a different set of parameters, only QDM accepts `max_bin`, a single parameter `as_quantile_dmatrix` might not be sufficient to reflect and clarify the difference, and my preference is still to have two different classes for them.

On the other hand, `ExternalMemDMatrix` is quite similar to `DMatrix` with the same set of parameters, and can be used "almost" all the places where DMatrix can be used.

The question is how to integrate the concept of external memory into the mix. I propose the following functions and classes, `ExternalMemoryDMatrix` as implemented in this PR, but without the support for being a QDM. In addition, a QDM constructor that accepts only the iterator `xgb.QuantileDMatrix.from_iterator(iter) -> xgb.QuantileDMatrix`.

The reasoning behind this is that the `xgb.QuantileDMatrix` and `xgb.QuantileDMatrix.from_iterator` return the same type with the exact same underlying data (in c++), it's just how it's constructed in the first place and so there's no need for a new class in R. Then we use different constructor to represent the difference of construction process. As for the external memory version of DMatrix, as you have suggested, they behave differently, we will continue with your approach of creating a new class for it.

What do you think?

Reply:
Sounds good.

So that'd leave the following functions with the following output classes:
* `xgb.DMatrix` -> `xgb.DMatrix`
* `xgb.QuantileDMatrix` -> `xgb.QuantileDMatrix`
* `xgb.QuantileDMatrix.from_iterator` -> `xgb.QuantileDMatrix`
* `xgb.ExternalDMatrix` -> `xgb.ExternalDMatrix`

Is that a correct understanding?

> I think as long as a missing value is accepted, integer will be considered as float.

Not quite - integer matrices are also accepted, and are passed as such to the array_interface parser. In such case, the missing value is auto-adjusted in the C++ code according to the type of the input, but if the user passes a data iterator, the missing value is passed before the iterator to the first C function, and thus cannot be converted later on once the data is seen and the dtype is determined (with integer missing value in R being `-INT_MAX`).

Reply:
> Is that a correct understanding?

Yes.

> and thus cannot be converted later on once the data is seen and the dtype is determined

That's interesting, did not think of that.

Reply:
Renamed.